File: examples/emotion/cpp/emotion.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * @file emotion.cpp
 * @brief Emotion recognition example using vision_service API.
 */

#include <iostream>    // NOLINT(build/include_order)
#include <memory>      // NOLINT(build/include_order)
#include <sstream>     // NOLINT(build/include_order)
#include <string>      // NOLINT(build/include_order)
#include <vector>      // NOLINT(build/include_order)
#include <fstream>     // NOLINT(build/include_order)
#include <iomanip>     // NOLINT(build/include_order)
#include <filesystem>  // NOLINT(build/c++17) NOLINT(build/include_order)

#include <opencv2/opencv.hpp>  // NOLINT(build/include_order)
#include <yaml-cpp/yaml.h>     // NOLINT(build/include_order)

#include "vision_service.h"  // NOLINT(build/include_order)

namespace fs = std::filesystem;

static std::string resolve_path(const fs::path& root, const std::string& path) {
    if (path.empty()) return path;
    if (path[0] == '/' || (path.size() >= 2 && path[1] == ':')) return path;
    return (root / path).lexically_normal().string();
}

std::vector<std::string> load_labels(const std::string& label_file_path) {
    std::vector<std::string> labels;
    std::ifstream file(label_file_path);
    if (!file.is_open()) return labels;
    std::string line;
    while (std::getline(file, line)) {
        if (!line.empty()) labels.push_back(line);
    }
    return labels;
}

void print_usage(const char* program_name) {
    std::cout << "Usage: " << program_name << " <config_yaml> [options]\n"
                << "  Emotion 情绪识别示例（vision_service 接口）\n"
                << "Options (any order after config_yaml):\n"
                << "  --model-path <path>   Override model_path in yaml\n"
                << "  --image <path>   Input image path\n"
                << "  --output <path>  Output image path (default: result_emotion.jpg)\n"
                << "  --help           Show this help\n"
                << "\nExample:\n"
                << "  " << program_name << " examples/emotion/config/emotion.yaml\n"
                << "  " << program_name
                << " examples/emotion/config/emotion.yaml --image face.png --output result_emotion.jpg\n";
}

int main(int argc, char* argv[]) {
    if (argc < 2 || std::string(argv[1]) == "--help") {
        print_usage(argv[0]);
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string output_path = "result_emotion.jpg";
    std::string model_path_override;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--help") {
            print_usage(argv[0]);
            return 0;
        } else if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--output" && i + 1 < argc) {
            output_path = argv[++i];
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    std::unique_ptr<VisionService> service = VisionService::Create(
        config_path,
        model_path_override,
        true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (image_path.empty()) {
        const std::string default_image = service->GetDefaultImage();
        if (!default_image.empty()) {
            image_path = default_image;
        }
    }
    if (image_path.empty()) {
        std::cerr << "Error: No --image and no test_image in config" << std::endl;
        return 1;
    }

    std::vector<std::string> labels;
    if (fs::exists(config_path)) {
        try {
            YAML::Node config = YAML::LoadFile(config_path);
            fs::path root = fs::path(config_path).parent_path().parent_path().parent_path().parent_path();
            if (config["label_file_path"]) {
                std::string lp = resolve_path(root, config["label_file_path"].as<std::string>());
                labels = load_labels(lp);
            }
        } catch (...) {}
    }
    if (labels.empty()) labels = {"neutral", "happy", "sad", "angry", "fear", "disgust", "surprise"};

    std::cout << "Loading image: " << image_path << std::endl;
    cv::Mat img = cv::imread(image_path);
    if (img.empty()) {
        std::cerr << "Error: Could not load image: " << image_path << std::endl;
        return 1;
    }
    std::vector<VisionServiceResult> results;
    VisionServiceStatus ret = service->InferImage(img, &results);
    if (ret != VISION_SERVICE_OK) {
        std::cerr << "Error: " << service->LastError() << std::endl;
        return 1;
    }

    if (results.empty()) {
        std::cout << "No emotion result." << std::endl;
        return 0;
    }

    int emotion_class = results[0].label;
    float emotion_score = results[0].score;
    std::string emotion_name = (emotion_class >= 0 && static_cast<size_t>(emotion_class) < labels.size())
                                ? labels[static_cast<size_t>(emotion_class)] : "unknown";

    std::cout << "Emotion: " << emotion_name << " (class " << emotion_class << ", score: "
                << std::fixed << std::setprecision(4) << emotion_score << ")" << std::endl;

    std::ostringstream oss;
    oss << emotion_name << " " << std::fixed << std::setprecision(2) << emotion_score;
    cv::putText(img, oss.str(), cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);
    cv::imwrite(output_path, img);
    std::cout << "Result saved to: " << output_path << std::endl;

    return 0;
}

File: examples/emotion/python/emotion.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
Emotion recognition example using spacemit_vision.

运行方式：通过 --config 指定 yaml 路径
→ VisionServiceNative.create(config_path) 创建模型
→ 从 yaml 读 test_image、label_file_path 等参数做推理。
"""

import argparse
from pathlib import Path
import cv2
import numpy as np
import yaml

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def load_labels(label_file: str) -> list:
    """读取标签文件：每行一个类别名。"""
    labels = []
    with open(label_file, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                labels.append(line)
    return labels


def parse_args():
    parser = argparse.ArgumentParser(description="Emotion Recognition Example")
    parser.add_argument(
        "--config",
        type=str,
        default=None,
        help="Config yaml path (default: examples/emotion/config/emotion.yaml)",
    )
    parser.add_argument(
        "--image",
        type=str,
        help="Input image path (if not provided, uses config default)",
    )
    parser.add_argument(
        "--output",
        type=str,
        default="result_emotion.jpg",
        help="Output image path",
    )
    parser.add_argument(
        "--model-path",
        type=str,
        default=None,
        help="Override model_path in yaml",
    )
    return parser.parse_args()


def main():
    args = parse_args()

    try:
        default_config = Path(__file__).parent.parent / "config" / "emotion.yaml"
        config_path = Path(args.config) if args.config else default_config
        project_root = Path(__file__).parent.parent.parent.parent
        model_name = config_path.stem

        print(f"创建 {model_name} 情绪识别器...")
        model_path_override = ""
        if args.model_path:
            p = Path(args.model_path).expanduser()
            model_path_override = str(
                p if p.is_absolute() else (project_root / p).resolve()
            )
        recognizer = VisionServiceNative.create(
            str(config_path), model_path_override=model_path_override
        )

        # Load config from specified yaml
        config = {}
        if config_path.exists():
            with open(config_path, "r", encoding="utf-8") as f:
                config = yaml.safe_load(f) or {}

        # Determine image path
        image_path = args.image or config.get("test_image", "~/.cache/assets/image/003_face0.png")
        image_path = resolve_path(image_path, project_root)

        print(f"加载图像: {image_path}")
        image = cv2.imread(str(image_path))
        if image is None:
            raise FileNotFoundError(f"无法加载图像: {image_path}")

        print(f"图像尺寸: {image.shape}")

        # Load labels
        labels = None
        label_file_path = config.get("label_file_path")
        if label_file_path:
            label_file_path = resolve_path(label_file_path, project_root)
            try:
                labels = load_labels(str(label_file_path))
                print(f"加载标签文件: {label_file_path} ({len(labels)} 个标签)")
            except Exception as e:
                print(f"警告: 无法加载标签文件 {label_file_path}: {e}")

        # Use default labels if not loaded
        if not labels:
            labels = [
                "neutral", "happiness", "sadness", "surprise",
                "fear", "disgust", "anger",
            ]

        print("运行情绪识别...")
        status, results = recognizer.infer_image(image)
        if status != VisionServiceStatus.OK:
            raise RuntimeError(recognizer.last_error())

        if not results:
            print("未得到情绪识别结果。")
            return 0

        scores = list(results[0].class_scores)
        if not scores:
            print("情绪识别结果的类别分数为空。")
            return 0

        emotion_class = int(np.argmax(scores))
        emotion_score = float(scores[emotion_class])

        emotion_name = (
            labels[emotion_class]
            if 0 <= emotion_class < len(labels)
            else "unknown"
        )
        print(f"Emotion: {emotion_name} (class {emotion_class}, score: {emotion_score:.4f})")

        result_image = image.copy()
        label_text = f"{emotion_name} {emotion_score:.2f}"
        cv2.putText(
            result_image,
            label_text,
            (10, 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            1.0,
            (0, 255, 0),
            2,
        )
        cv2.imwrite(args.output, result_image)
        print(f"结果已保存到: {args.output}")

    except Exception as e:
        print(f"错误: {e}")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())
